advanced_source/ddp_pipeline_tutorial.py

```diff
@@ -0,0 +1,452 @@
+# with the ``S`` dimension in the Transformer model. The batch dimension
+# ``N`` is along dimension 1.
+#
+
+# In 'run_worker'
+    bptt = 35
+    def get_batch(source, i):
+        seq_len = min(bptt, len(source) - 1 - i)
+        data = source[i:i+seq_len]
+        target = source[i+1:i+1+seq_len].view(-1)
+        return data, target
+
+######################################################################
+# Model scale and Pipe initialization
+# -----------------------------------
+#
+
+
+######################################################################
+# To demonstrate training large Transformer models using pipeline parallelism,
+# we scale up the Transformer layers appropriately. We use an embedding
+# dimension of 4096, hidden size of 4096, 16 attention heads and 8 total
+# transformer layers (``nn.TransformerEncoderLayer``). This creates a model with
+# **~1 billion** parameters.
+#
+# We need to initialize the `RPC Framework <https://pytorch.org/docs/stable/rpc.html>`__
+# since Pipe depends on the RPC framework via `RRef <https://pytorch.org/docs/stable/rpc.html#rref>`__
+# which allows for future expansion to cross host pipelining. We need to
+# initialize the RPC framework with only a single worker since we're using a
+# single process to drive multiple GPUs.
+#
+# The pipeline is then initialized with 8 transformer layers on one GPU and 8
+# transformer layers on the other GPU. One pipe is setup across GPUs 0 and 1 and
+# another across GPUs 2 and 3. Both pipes are then replicated using DistributedDataParallel.
+
+# In 'run_worker'
+    ntokens = len(TEXT.vocab.stoi) # the size of vocabulary
+    emsize = 4096 # embedding dimension
+    nhid = 4096 # the dimension of the feedforward network model in nn.TransformerEncoder
+    nlayers = 8 # the number of nn.TransformerEncoderLayer in nn.TransformerEncoder
+    nhead = 16 # the number of heads in the multiheadattention models
+    dropout = 0.2 # the dropout value
+
+    from torch.distributed import rpc
+    tmpfile = tempfile.NamedTemporaryFile()
+    rpc.init_rpc(
+        name="worker",
+        rank=0,
+        world_size=1,
+        rpc_backend_options=rpc.TensorPipeRpcBackendOptions(
+            init_method="file://{}".format(tmpfile.name),
+        )
+    )
+
+    # Need to use 'checkpoint=never' since as of PyTorch 1.8, Pipe checkpointing
+    # doesn't work with DDP.
+    from torch.distributed.pipeline.sync import Pipe
+    model = Pipe(
+                torch.nn.Sequential(
+                    TransformerModelStage1(ntokens, emsize, nhead, nhid, int(nlayers/2), dropout).cuda(2 * rank),
+                    TransformerModelStage2(ntokens, emsize, nhead, nhid, int(nlayers/2), dropout).cuda(2 * rank + 1),
+                ),
+                chunks = 8,
+                checkpoint = "never"
+            )
+
+    # Initialize process group and wrap model in DDP.
+    from torch.nn.parallel import DistributedDataParallel
```

Comment:
is there a technical reason we import here?

Reply:
Just closer to the usage of the module.